File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.py
# 1. 가장 먼곳부터 처리
# 2. 갈때는 배달만, 올때는 수거만
# 3. 항상 최대치를 배달과 수거

def solution(cap, n, deliveries, pickups):
    answer = 0
    deliver = []
    pickup = []
    for i, v in enumerate(deliveries):
        if v > 0:
            deliver.append((i+1, v))
    for i, v in enumerate(pickups):
        if v > 0:
            pickup.append((i+1, v))
    while deliver or pickup:
        distance = 0
        # 배달
        temp = cap
        while deliver and temp > 0:
            d, v = deliver.pop()
            distance = max(d, distance)
            temp -= v
        if temp < 0:
            deliver.append((d, -temp))
        # 수거
        temp = cap
        while pickup and temp > 0:
            d, v = pickup.pop()
            distance = max(d, distance)
            temp -= v
        if temp < 0:
            pickup.append((d, -temp))
        # 거리 더하기
        answer += distance * 2
        
    return answer

File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.js
function solution(cap, n, deliveries, pickups) {
    var answer = 0;
    let maxDelivery, maxPickup;
    for (let i = n-1; i > -1; i--) {
        if (deliveries[i] > 0) {
            maxDelivery = i;
            break;
        }
    }
    for (let i = n-1; i > -1; i--) {
        if (pickups[i] > 0) {
            maxPickup = i;
            break;
        }
    }
    
    let distance, spare, idx;
    while (maxDelivery >= 0 || maxPickup >= 0) {
        answer += Math.max(maxDelivery+1, maxPickup+1) * 2;
        // 배달
        spare = cap;
        idx = maxDelivery;
        while (spare >= 0 && idx >= 0) {
            deliveries[idx] -= spare;
            spare = -1 * deliveries[idx];
            idx -= 1;
            if (spare >= 0) {
                maxDelivery = idx;
            }
        }
        // 수거
        spare = cap;
        idx = maxPickup;
        while (spare >= 0 && idx >= 0) {
            pickups[idx] -= spare;
            spare = -1 * pickups[idx];
            idx -= 1;
            if (spare >= 0) {
                maxPickup = idx;
            }
        }
        
    }
    
    return answer;
}